qiskit/circuit/quantumcircuit.py

```diff
@@ -3044,3 +2979,81 @@ def _compare_parameters(param1, param2):
 
     # else sort by name
     return _standard_compare(param1.name, param2.name)
+
+
+def _get_composite_circuit_qasm_from_instruction(
+    instruction, existing_gate_names, existing_composite_circuits
+):
+    """Returns OpenQASM string composite circuit given an instruction.
+    The given instruction should be the result of composite_circuit.to_instruction()."""
+
+    if instruction.definition is None:
+        raise ValueError(f'Instruction "{instruction.name}" is not defined.')
+
+    gate_parameters = ",".join(["param%i" % num for num in range(len(instruction.params))])
+    qubit_parameters = ",".join(["q%i" % num for num in range(instruction.num_qubits)])
+    composite_circuit_gates = ""
+
+    definition = instruction.definition
+    definition_bit_labels = {
+        bit: idx for bits in (definition.qubits, definition.clbits) for idx, bit in enumerate(bits)
+    }
+    for sub_instruction, qargs, _ in definition:
+        if sub_instruction.name not in existing_gate_names:
+            if sub_instruction not in existing_composite_circuits:
+                existing_composite_circuits.insert(0, sub_instruction)
+                _get_composite_circuit_qasm_from_instruction(
+                    sub_instruction, existing_gate_names, existing_composite_circuits
+                )
+        gate_qargs = ",".join(
+            ["q%i" % index for index in [definition_bit_labels[qubit] for qubit in qargs]]
+        )
+        composite_circuit_gates += "%s %s; " % (sub_instruction.qasm(), gate_qargs)
+
+    if composite_circuit_gates:
+        composite_circuit_gates = composite_circuit_gates.rstrip(" ")
+
+    if gate_parameters:
+        qasm_string = "gate %s(%s) %s { %s }" % (
+            instruction.name,
+            gate_parameters,
+            qubit_parameters,
+            composite_circuit_gates,
+        )
+    else:
+        qasm_string = "gate %s %s { %s }" % (
+            instruction.name,
+            qubit_parameters,
+            composite_circuit_gates,
+        )
+
+    return qasm_string
+
+
+def _insert_composite_gate_definition_qasm(
+    string_temp, existing_gate_names, existing_composite_circuits, extension_lib
+):
+    """Insert composite gate definition QASM code right after extension library in the header"""
+
+    gate_definition_string = ""
+
+    # Cycle through all gate definitions and add all undefined gates to the list
+    for instruction in existing_composite_circuits:
+        _get_composite_circuit_qasm_from_instruction(
+            instruction, existing_gate_names, existing_composite_circuits
+        )
+
+    # Generate gate definition string
+    for instruction in existing_composite_circuits:
+        if hasattr(instruction, "_qasm_definition"):
```

Comment:
Nothing to change, just for me to understand: only the `UnitaryGate` has the `_qasm_definition` attribute, right?

Reply:
I think so. This part was modified by @1ucian0. My original code was checking the instance of `UnitaryGate`: https://github.com/Qiskit/qiskit-terra/pull/5887/commits/a9e80dcab2b6af4c920aa6ae1ecb5d7f0c54b9b8

Reply:
Correct. Maybe we should have a standard way for a `Gate` to express "here is my QASM definition".